File: worklist-ui/src/main/java/com/gcgenome/lims/client/domain/Request.java
package com.gcgenome.lims.client.domain;

import com.google.gwt.i18n.client.DateTimeFormat;
import jsinterop.annotations.*;
import lombok.Getter;
import lombok.experimental.Accessors;

import java.util.Date;

@JsType(isNative=true, namespace= JsPackage.GLOBAL, name="Object")
@Getter(onMethod_= {@JsOverlay, @JsIgnore})
@Accessors(fluent=true)
public final class Request {
    private Sample sample;
    private Service service;
    private Organization requester;
    @JsProperty(name="date_request")
    private Double dateRequest;
    @JsProperty(name="date_reception")
    private Double dateReception;
    @JsProperty(name="date_due")
    private Double dateDue;
    @JsProperty(name="date_due_publish")
    private Double dateDuePublish;
    private String remark;
    @JsOverlay @JsIgnore
    public String id() {
        try {
            return sample.id() + "$" + service.id();
        } catch(Exception e) {
            return null;
        }
    }
    @JsOverlay @JsIgnore
    private final static DateTimeFormat DEFAULT_DATE_FORMAT = DateTimeFormat.getFormat("yyyy-MM-dd");
    @JsOverlay @JsIgnore
    public String dateRequest() {
        if(dateRequest==null) return null;
        return DEFAULT_DATE_FORMAT.format(new Date(dateRequest.longValue()));
    }
    @JsOverlay @JsIgnore
    public String dateReception() {
        if(dateReception==null) return null;
        return DEFAULT_DATE_FORMAT.format(new Date(dateReception.longValue()));
    }
    @JsOverlay @JsIgnore
    public String dateDue() {
        if(dateDue==null) return null;
        return DEFAULT_DATE_FORMAT.format(new Date(dateDue.longValue()));
    }
    @JsOverlay @JsIgnore
    public String dateDuePublish() {
        if(dateDuePublish==null) return null;
        return DEFAULT_DATE_FORMAT.format(new Date(dateDuePublish.longValue()));
    }
}

File: domain/src/main/kotlin/com/gcgenome/lims/domain/Request.kt
package com.gcgenome.lims.domain

import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter

@JvmRecord
data class Request (
    val sample: Sample,
    val service: Service,
    val requester: Organization,
    val dateRequest: Long?,
    val dateReception: Long?,
    val dateDue: Long?,
    val dateDuePublish: Long?,
    val barcode: String? // 의뢰기관 바코드(Optional)
) {
    constructor(
        sample: Sample,
        service: Service,
        requester: Organization,
        dateRequest: LocalDateTime?,
        dateReception: LocalDateTime?,
        dateDue: LocalDateTime?,
        dateDuePublish: LocalDateTime?,
        barcode: String? = null
    ): this (
        sample=sample,
        service=service,
        requester=requester,
        dateRequest=dateRequest.toEpochMilli(),
        dateReception=dateReception.toEpochMilli(),
        dateDue=dateDue.toEpochMilli(),
        dateDuePublish=dateDuePublish.toEpochMilli(),
        barcode=barcode
    )
    companion object {
        val DF: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
        fun LocalDateTime?.toEpochMilli(zoneId: ZoneId = ZoneId.systemDefault()): Long? =
        this?.atZone(zoneId)?.toInstant()?.toEpochMilli()
    }
}